mscore/plugin/pluginUpdater.cpp

```diff
@@ -0,0 +1,866 @@
+      while (it.hasNext()) {
+            QRegularExpressionMatch match = it.next();
+            PluginPackageLink link;
+            link.source = GITHUB_RELEASE;
+            link.url = match.captured(1);
+            link.user = match.captured(2);
+            link.repo = match.captured(3);
+            localHint(match, link, html_raw);
+            github_urls.push_back(link);
+            }
+      it = github_directory_patt.globalMatch(html_raw);
+      while (it.hasNext()) {
+            QRegularExpressionMatch match = it.next();
+            PluginPackageLink link;
+            link.source = GITHUB;
+            link.url = match.captured(1);
+            link.user = match.captured(2);
+            link.repo = match.captured(3);
+            link.branch = match.captured(5);
+            localHint(match, link, html_raw);
+            github_urls.push_back(link);
+            }
+      it = github_repo_patt.globalMatch(html_raw);
+      while (it.hasNext()) {
+            QRegularExpressionMatch match = it.next();
+            PluginPackageLink link;
+            link.source = GITHUB;
+            link.url = match.captured(1);
+            link.user = match.captured(2);
+            link.repo = match.captured(3);
+            localHint(match, link, html_raw);
+            github_urls.push_back(link);
+            }
+      return github_urls;
+      }
+
+static std::vector<PluginPackageLink> getLinks(const QByteArray& html_raw_array)
+      {
+      QString html_raw(html_raw_array);
+      auto GetLocalHint = [](QRegularExpressionMatch& match, PluginPackageLink& link, const QString& html_raw) {
+            int start_idx = match.capturedStart(0);
+            int end_idx = match.capturedEnd();
+            int newline_start = start_idx, newline_end = end_idx;
+            while (newline_start > 0 && html_raw[newline_start - 1] != '\n')
+                  newline_start--;
+            while (newline_end < html_raw.length() && html_raw[newline_end] != '\n')
+                  newline_end++;
+            // TODO: first use hyper link text as hint, 
+            // and if not strong enough, use the whole line 
+            link.newline_index = newline_start;
+            link.curr_line = html_raw.mid(newline_start, newline_end - newline_start + 1);
+            link.score = CompatEstimate(link.curr_line);
+            };
+      std::vector<PluginPackageLink> attachments = getAttachments(html_raw, GetLocalHint);
+      std::vector<PluginPackageLink> github_links = getGitHubLinks(html_raw, GetLocalHint);
+      std::vector<PluginPackageLink> urls;
+      urls.insert(urls.begin(), attachments.begin(), attachments.end());
+      urls.insert(urls.begin(), github_links.begin(), github_links.end());
+      std::set<int> begin_idx;
+      for (auto& url : urls)
+            begin_idx.insert(url.newline_index);
+      // now, get more hints from the line above each link
+      for (int i = 0; i < urls.size(); i++) {
```

Comment:
compiler warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
Guess making `i`  a `site_t` would fix it.